src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs

```diff
@@ -123,6 +136,24 @@ public async Task StartAsync(CancellationToken cancellationToken)
         }
 
         private async Task StartHostAsync(CancellationToken cancellationToken, int attemptCount = 0, JobHostStartupMode startupMode = JobHostStartupMode.Normal)
+        {
+            try
+            {
+                await _hostStartSemaphore.WaitAsync();
+                await StartHostUnlockedAsync(cancellationToken, attemptCount, startupMode);
+            }
+            finally
+            {
+                _hostStartSemaphore.Release();
+            }
+        }
+
+        /// <summary>
+        /// Starts the host without taking a lock. Callers must take a lock on _hostStartSemaphore
+        /// before calling this method. Host starts and restarts must be synchronous to prevent orphaned
+        /// hosts or an incorrect ActiveHost.
+        /// </summary>
+        private async Task StartHostUnlockedAsync(CancellationToken cancellationToken, int attemptCount = 0, JobHostStartupMode startupMode = JobHostStartupMode.Normal)
```

Comment:
I wanted to name this something that made any callers think about it -- open to suggestions though. Using advice from here: https://stackoverflow.com/a/19825074/3516125.